File: client/src/services/patientService.ts
import api from "./api";
import type { ApiResponse, Patient, PagedResponse } from "../types";

export interface DoctorPatient {
    id: string;
    fullName: string;
    email: string;
    phone: string;
    profileImage: string;
    gender: string;
    bloodGroup: string;
    allergies: string;
    medicalConditions: string;
    dateOfBirth: string;
    totalVisits: number;
    lastVisitDate: string;
    status: "ACTIVE" | "PAST";
    visits?: VisitSummary[];
}

export interface VisitSummary {
    appointmentId: string;
    appointmentDate: string;
    slotTime: string;
    consultationType: string;
    status: string;
    reason: string;
    diagnosis: string;
    prescriptions: PrescriptionSummary[];
}

export interface PrescriptionSummary {
    medicationName: string;
    dosage: string;
    frequency: string;
    duration: string;
}

export interface PatientProfileRequest {
    firstName: string;
    lastName: string;
    dateOfBirth: string;
    gender: "MALE" | "FEMALE" | "OTHER";
    bloodGroup: string;
    address: string;
    emergencyContact: string;
    allergies?: string;
    medicalConditions?: string;
}

export const patientService = {
    async getAllPatients(
        page = 0,
        size = 10,
        sortBy = "createdAt",
        sortDir = "desc"
    ): Promise<PagedResponse<Patient>> {
        const response = await api.get<ApiResponse<PagedResponse<Patient>>>(
            `/patients?page=${page}&size=${size}&sortBy=${sortBy}&sortDir=${sortDir}`
        );
        return response.data.data;
    },

    async searchPatients(
        query: string,
        page = 0,
        size = 10
    ): Promise<PagedResponse<Patient>> {
        const response = await api.get<ApiResponse<PagedResponse<Patient>>>(
            `/patients/search?query=${query}&page=${page}&size=${size}`
        );
        return response.data.data;
    },

    async getOwnProfile(): Promise<Patient> {
        const response = await api.get<ApiResponse<Patient>>("/patients/profile");
        return response.data.data;
    },

    async updateProfile(data: PatientProfileRequest): Promise<Patient> {
        const response = await api.put<ApiResponse<Patient>>("/patients/profile", data);
        return response.data.data;
    },

    async getPatientById(id: string): Promise<Patient> {
        const response = await api.get<ApiResponse<Patient>>(`/patients/${id}`);
        return response.data.data;
    },

    async getDoctorPatients(): Promise<DoctorPatient[]> {
        const response = await api.get<ApiResponse<DoctorPatient[]>>(
            "/appointments/my-patients"
        );
        return response.data.data;
    },

    async getDoctorPatientDetail(patientId: string): Promise<DoctorPatient> {
        const response = await api.get<ApiResponse<DoctorPatient>>(
            `/appointments/my-patients/${patientId}`
        );
        return response.data.data;
    },
};

File: server/src/main/java/com/doconsult/server/service/PatientService.java
package com.doconsult.server.service;

import com.doconsult.server.dto.request.PatientProfileRequest;
import com.doconsult.server.dto.response.PagedResponse;
import com.doconsult.server.dto.response.PatientResponse;
import com.doconsult.server.exception.ResourceNotFoundException;
import com.doconsult.server.model.Patient;
import com.doconsult.server.repository.PatientRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class PatientService {

        private final PatientRepository patientRepository;

        public PagedResponse<PatientResponse> getAllPatients(int page, int size, String sortBy, String sortDir) {
                Sort sort = sortDir.equalsIgnoreCase("desc")
                                ? Sort.by(sortBy).descending()
                                : Sort.by(sortBy).ascending();
                Pageable pageable = PageRequest.of(page, size, sort);

                Page<Patient> patientsPage = patientRepository.findAll(pageable);

                List<PatientResponse> content = patientsPage.getContent().stream()
                                .map(this::mapToResponse)
                                .collect(Collectors.toList());

                return PagedResponse.<PatientResponse>builder()
                                .content(content)
                                .page(patientsPage.getNumber())
                                .size(patientsPage.getSize())
                                .totalElements(patientsPage.getTotalElements())
                                .totalPages(patientsPage.getTotalPages())
                                .last(patientsPage.isLast())
                                .first(patientsPage.isFirst())
                                .build();
        }

        public PagedResponse<PatientResponse> searchPatients(String search, int page, int size) {
                Pageable pageable = PageRequest.of(page, size);
                Page<Patient> patientsPage = patientRepository.searchPatients(search, pageable);

                List<PatientResponse> content = patientsPage.getContent().stream()
                                .map(this::mapToResponse)
                                .collect(Collectors.toList());

                return PagedResponse.<PatientResponse>builder()
                                .content(content)
                                .page(patientsPage.getNumber())
                                .size(patientsPage.getSize())
                                .totalElements(patientsPage.getTotalElements())
                                .totalPages(patientsPage.getTotalPages())
                                .last(patientsPage.isLast())
                                .first(patientsPage.isFirst())
                                .build();
        }

        public PatientResponse getPatientById(java.util.UUID id) {
                Patient patient = patientRepository.findById(id)
                                .orElseThrow(() -> new ResourceNotFoundException("Patient", "id", id));
                return mapToResponse(patient);
        }

        public PatientResponse getPatientByUserId(java.util.UUID userId) {
                Patient patient = patientRepository.findByUserId(userId)
                                .orElseThrow(() -> new ResourceNotFoundException("Patient profile", "userId", userId));
                return mapToResponse(patient);
        }

        @Transactional
        public PatientResponse createOrUpdateProfile(java.util.UUID userId, PatientProfileRequest request) {
                Patient patient = patientRepository.findByUserId(userId)
                                .orElseThrow(() -> new ResourceNotFoundException("Patient profile", "userId", userId));

                patient.setFirstName(request.getFirstName());
                patient.setLastName(request.getLastName());
                patient.setDateOfBirth(request.getDateOfBirth());
                patient.setGender(request.getGender());
                patient.setBloodGroup(request.getBloodGroup());
                patient.setAddress(request.getAddress());
                patient.setEmergencyContact(request.getEmergencyContact());
                patient.setAllergies(request.getAllergies());
                patient.setMedicalConditions(request.getMedicalConditions());

                patient = patientRepository.save(patient);
                return mapToResponse(patient);
        }

        private PatientResponse mapToResponse(Patient patient) {
                return PatientResponse.builder()
                                .id(patient.getId())
                                .userId(patient.getUser().getId())
                                .email(patient.getUser().getEmail())
                                .phone(patient.getUser().getPhone())
                                .firstName(patient.getFirstName())
                                .lastName(patient.getLastName())
                                .fullName(patient.getFullName())
                                .dateOfBirth(patient.getDateOfBirth())
                                .gender(patient.getGender())
                                .bloodGroup(patient.getBloodGroup())
                                .address(patient.getAddress())
                                .emergencyContact(patient.getEmergencyContact())
                                .profileImage(patient.getProfileImage())
                                .allergies(patient.getAllergies())
                                .medicalConditions(patient.getMedicalConditions())
                                .build();
        }
}
